File: hphp/hack/src/hhbc/emit_memoize_function.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Core_kernel
open Instruction_sequence
open Emit_memoize_helpers
open Hhbc_ast
module R = Hhbc_string_utils.Reified
module T = Aast

let make_memoize_function_no_params_code
    ~deprecation_info env renamed_function_id is_async =
  let notfound = Label.next_regular () in
  let suspended_get = Label.next_regular () in
  let eager_set = Label.next_regular () in
  let scope = Emit_env.get_scope env in
  let deprecation_body =
    Emit_body.emit_deprecation_warning scope deprecation_info
  in
  let fcall_args =
    if is_async then
      make_fcall_args ~async_eager_label:eager_set 0
    else
      make_fcall_args 0
  in
  gather
    [
      deprecation_body;
      ( if is_async then
        gather
          [
            instr_memoget_eager notfound suspended_get None;
            instr_retc;
            instr_label suspended_get;
            instr_retc_suspended;
          ]
      else
        gather [instr_memoget notfound None; instr_retc] );
      instr_label notfound;
      instr_nulluninit;
      instr_nulluninit;
      instr_nulluninit;
      instr_fcallfuncd fcall_args renamed_function_id;
      instr_memoset None;
      ( if is_async then
        gather
          [
            instr_retc_suspended;
            instr_label eager_set;
            instr_memoset_eager None;
            instr_retc;
          ]
      else
        gather [instr_retc] );
    ]

let make_memoize_function_with_params_code
    ~pos
    ~deprecation_info
    env
    params
    ast_params
    renamed_method_id
    is_async
    is_reified =
  let param_count = List.length params in
  let notfound = Label.next_regular () in
  let suspended_get = Label.next_regular () in
  let eager_set = Label.next_regular () in
  (* The local that contains the reified generics is the first non parameter
   * local, so the first local is parameter count + 1 when there are reified =
   * generics *)
  let add_reified =
    if is_reified then
      1
    else
      0
  in
  let first_local = Local.Unnamed (param_count + add_reified) in
  let (begin_label, default_value_setters) =
    (* Default value setters belong in the wrapper function not in the original function *)
    Emit_param.emit_param_default_value_setter env pos params
  in
  let scope = Emit_env.get_scope env in
  let deprecation_body =
    Emit_body.emit_deprecation_warning scope deprecation_info
  in
  let fcall_args =
    let flags = { default_fcall_flags with has_generics = is_reified } in
    if is_async then
      make_fcall_args ~flags ~async_eager_label:eager_set param_count
    else
      make_fcall_args ~flags param_count
  in
  let (reified_get, reified_memokeym) =
    if not is_reified then
      (empty, empty)
    else
      ( instr_cgetl (Local.Named R.reified_generics_local_name),
        gather
        @@ getmemokeyl
             param_count
             (param_count + add_reified)
             R.reified_generics_local_name )
  in
  let param_count = param_count + add_reified in
  gather
    [
      begin_label;
      Emit_body.emit_method_prolog
        ~env
        ~pos
        ~params
        ~ast_params
        ~tparams:[]
        ~should_emit_init_this:false;
      deprecation_body;
      param_code_sets params param_count;
      reified_memokeym;
      ( if is_async then
        gather
          [
            instr_memoget_eager
              notfound
              suspended_get
              (Some (first_local, param_count));
            instr_retc;
            instr_label suspended_get;
            instr_retc_suspended;
          ]
      else
        gather
          [instr_memoget notfound (Some (first_local, param_count)); instr_retc]
      );
      instr_label notfound;
      instr_nulluninit;
      instr_nulluninit;
      instr_nulluninit;
      param_code_gets params;
      reified_get;
      instr_fcallfuncd fcall_args renamed_method_id;
      instr_memoset (Some (first_local, param_count));
      ( if is_async then
        gather
          [
            instr_retc_suspended;
            instr_label eager_set;
            instr_memoset_eager (Some (first_local, param_count));
            instr_retc;
          ]
      else
        gather [instr_retc] );
      default_value_setters;
    ]

let make_memoize_function_code
    ~pos
    ~deprecation_info
    env
    params
    ast_params
    renamed_method_id
    is_async
    is_reified =
  let fun_ =
    if List.is_empty params && not is_reified then
      make_memoize_function_no_params_code
        ~deprecation_info
        env
        renamed_method_id
        is_async
    else
      make_memoize_function_with_params_code
        ~pos
        ~deprecation_info
        env
        params
        ast_params
        renamed_method_id
        is_async
        is_reified
  in
  Emit_pos.emit_pos_then pos fun_

(* Construct the wrapper function body *)
let make_wrapper_body env return_type params instrs is_reified =
  Emit_body.make_body
    instrs
    ( if is_reified then
      [R.reified_generics_local_name]
    else
      []
    (* decl_vars *) )
    true (* is_memoize_wrapper *)
    false (* is_memoize_wrapper_lsb *)
    [] (* upper_bounds *)
    [] (* shadowed_tparams *)
    params
    (Some return_type)
    None (* doc *)
    (Some env)

let emit_wrapper_function
    ~original_id ~renamed_id ~is_method ~deprecation_info ast_fun =
  Emit_memoize_helpers.check_memoize_possible
    (fst ast_fun.T.f_name)
    ~params:ast_fun.T.f_params
    ~is_method;
  let scope = [Ast_scope.ScopeItem.Function ast_fun] in
  let namespace = ast_fun.T.f_namespace in
  let pos = ast_fun.T.f_span in
  let env = Emit_env.(empty |> with_namespace namespace |> with_scope scope) in
  let tparams =
    List.map (Ast_scope.Scope.get_tparams scope) (fun t -> snd t.T.tp_name)
  in
  let params =
    Emit_param.from_asts
      ~namespace
      ~tparams
      ~generate_defaults:true
      ~scope
      ast_fun.T.f_params
  in
  let function_attributes =
    Emit_attribute.from_asts namespace ast_fun.T.f_user_attributes
  in
  let function_attributes =
    Emit_attribute.add_reified_attribute function_attributes ast_fun.T.f_tparams
  in
  let function_is_async = ast_fun.T.f_fun_kind = Ast_defs.FAsync in
  let scope = [Ast_scope.ScopeItem.Function ast_fun] in
  let return_type_info =
    Emit_body.emit_return_type_info
      ~scope
      ~skipawaitable:function_is_async
      (T.hint_of_type_hint ast_fun.T.f_ret)
  in
  let is_reified =
    List.exists
      ~f:(fun t -> t.T.tp_reified = T.Reified || t.T.tp_reified = T.SoftReified)
      ast_fun.T.f_tparams
  in
  let body_instrs =
    make_memoize_function_code
      ~pos
      ~deprecation_info
      env
      params
      ast_fun.T.f_params
      renamed_id
      function_is_async
      is_reified
  in
  let function_rx_level =
    Rx.rx_level_from_ast ast_fun.T.f_user_attributes
    |> Option.value ~default:Rx.NonRx
  in
  let env = Emit_env.with_rx_body (function_rx_level <> Rx.NonRx) env in
  let memoized_body =
    make_wrapper_body env return_type_info params body_instrs is_reified
  in
  let is_interceptable = Interceptable.is_function_interceptable ast_fun in
  Hhas_function.make
    function_attributes
    original_id
    memoized_body
    (Hhas_pos.pos_to_span ast_fun.T.f_span)
    function_is_async
    false (* is_generator *)
    false (* is_pair_generator *)
    Closure_convert.TopLevel
    false (* no_injection *)
    is_interceptable
    false (* is_memoize_impl *)
    function_rx_level
    (* function_rx_disabled *)
    false

File: hphp/hack/src/hhbc/emit_memoize_function.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use ast_scope_rust::{self as ast_scope, Scope, ScopeItem};
use emit_attribute_rust as emit_attribute;
use emit_body_rust as emit_body;
use emit_memoize_helpers_rust as emit_memoize_helpers;
use emit_param_rust as emit_param;
use emit_pos_rust::emit_pos_then;
use env::{emitter::Emitter, Env};
use hhas_body_rust::HhasBody;
use hhas_coeffects::HhasCoeffects;
use hhas_function_rust::{Flags as HhasFunctionFlags, HhasFunction};
use hhas_param_rust::HhasParam;
use hhas_pos_rust::Span;
use hhas_type::Info as HhasTypeInfo;
use hhbc_ast_rust::{FcallArgs, FcallFlags};
use hhbc_id_rust::{function::Type as FunId, Id};
use hhbc_string_utils_rust::{coeffects, reified};
use instruction_sequence::{instr, InstrSeq, Result};
use ocamlrep::rc::RcOc;
use options::{HhvmFlags, Options, RepoFlags};
use oxidized::{ast as T, pos::Pos};
use runtime::TypedValue;

pub(crate) fn is_interceptable(opts: &Options) -> bool {
    opts.hhvm
        .flags
        .contains(HhvmFlags::JIT_ENABLE_RENAME_FUNCTION)
        && !opts.repo_flags.contains(RepoFlags::AUTHORITATIVE)
}

pub(crate) fn emit_wrapper_function<'a>(
    emitter: &mut Emitter,
    original_id: FunId<'a>,
    renamed_id: &FunId,
    deprecation_info: &Option<&[TypedValue]>,
    f: &'a T::Fun_,
) -> Result<HhasFunction<'a>> {
    emit_memoize_helpers::check_memoize_possible(&(f.name).0, &f.params, false)?;
    let scope = Scope {
        items: vec![ScopeItem::Function(ast_scope::Fun::new_ref(f))],
    };
    let mut tparams = scope
        .get_tparams()
        .iter()
        .map(|tp| tp.name.1.as_str())
        .collect::<Vec<_>>();
    let params =
        emit_param::from_asts(emitter, &mut tparams, &f.namespace, true, &scope, &f.params)?;
    let mut attributes = emit_attribute::from_asts(emitter, &f.namespace, &f.user_attributes)?;
    attributes.extend(emit_attribute::add_reified_attribute(&f.tparams));
    let return_type_info = emit_body::emit_return_type_info(
        &tparams,
        f.fun_kind.is_fasync(), /* skip_awaitable */
        f.ret.1.as_ref(),
    )?;
    let is_reified = f
        .tparams
        .iter()
        .any(|tp| tp.reified.is_reified() || tp.reified.is_soft_reified());
    let mut env = Env::default(RcOc::clone(&f.namespace)).with_scope(scope);
    // TODO(hrust): avoid cloning
    let renamed_id_owned: FunId<'static> = renamed_id.to_raw_string().to_string().into();
    let body_instrs = make_memoize_function_code(
        emitter,
        &mut env,
        &f.span,
        *deprecation_info,
        &params,
        &f.params,
        renamed_id_owned,
        f.fun_kind.is_fasync(),
        is_reified,
    )?;
    let coeffects = HhasCoeffects::from_ast(&f.ctxs, &f.params);
    let has_coeffects_local = coeffects.has_coeffects_local();
    env.with_rx_body(coeffects.is_any_rx_or_pure());
    let body = make_wrapper_body(
        emitter,
        env,
        return_type_info,
        params,
        body_instrs,
        is_reified,
        has_coeffects_local,
    )?;

    let mut flags = HhasFunctionFlags::empty();
    flags.set(HhasFunctionFlags::ASYNC, f.fun_kind.is_fasync());
    let is_interceptable = is_interceptable(emitter.options());
    flags.set(HhasFunctionFlags::INTERCEPTABLE, is_interceptable);

    Ok(HhasFunction {
        attributes,
        name: original_id,
        body,
        span: Span::from_pos(&f.span),
        coeffects,
        flags,
    })
}

fn make_memoize_function_code(
    e: &mut Emitter,
    env: &mut Env,
    pos: &Pos,
    deprecation_info: Option<&[TypedValue]>,
    hhas_params: &[HhasParam],
    ast_params: &[T::FunParam],
    renamed_id: FunId<'static>,
    is_async: bool,
    is_reified: bool,
) -> Result {
    let fun = if hhas_params.is_empty() && !is_reified {
        make_memoize_function_no_params_code(e, env, deprecation_info, renamed_id, is_async)
    } else {
        make_memoize_function_with_params_code(
            e,
            env,
            pos,
            deprecation_info,
            hhas_params,
            ast_params,
            renamed_id,
            is_async,
            is_reified,
        )
    }?;
    Ok(emit_pos_then(pos, fun))
}

fn make_memoize_function_with_params_code(
    e: &mut Emitter,
    env: &mut Env,
    pos: &Pos,
    deprecation_info: Option<&[TypedValue]>,
    hhas_params: &[HhasParam],
    ast_params: &[T::FunParam],
    renamed_id: FunId<'static>,
    is_async: bool,
    is_reified: bool,
) -> Result {
    let param_count = hhas_params.len();
    let notfound = e.label_gen_mut().next_regular();
    let suspended_get = e.label_gen_mut().next_regular();
    let eager_set = e.label_gen_mut().next_regular();
    // The local that contains the reified generics is the first non parameter local,
    // so the first local is parameter count + 1 when there are reified = generics
    let add_refied = usize::from(is_reified);
    let first_local = local::Type::Unnamed(param_count + add_refied);
    let deprecation_body =
        emit_body::emit_deprecation_info(&env.scope, deprecation_info, e.systemlib())?;
    let (begin_label, default_value_setters) =
        // Default value setters belong in the wrapper method not in the original method
        emit_param::emit_param_default_value_setter(e, env, pos, hhas_params)?;
    let fcall_args = {
        let mut fcall_flags = FcallFlags::default();
        fcall_flags.set(FcallFlags::HAS_GENERICS, is_reified);
        FcallArgs::new(
            fcall_flags,
            1,
            vec![],
            if is_async {
                Some(eager_set.clone())
            } else {
                None
            },
            param_count,
            None,
        )
    };
    let (reified_get, reified_memokeym) = if !is_reified {
        (instr::empty(), instr::empty())
    } else {
        (
            instr::cgetl(local::Type::Named(reified::GENERICS_LOCAL_NAME.into())),
            InstrSeq::gather(emit_memoize_helpers::get_memo_key_list(
                param_count,
                param_count + add_refied,
                reified::GENERICS_LOCAL_NAME.into(),
            )),
        )
    };
    let param_count = (param_count + add_refied) as isize;
    Ok(InstrSeq::gather(vec![
        begin_label,
        emit_body::emit_method_prolog(e, env, pos, hhas_params, ast_params, &[])?,
        deprecation_body,
        emit_memoize_helpers::param_code_sets(hhas_params, param_count as usize),
        reified_memokeym,
        if is_async {
            InstrSeq::gather(vec![
                instr::memoget_eager(
                    notfound.clone(),
                    suspended_get.clone(),
                    Some((first_local.clone(), param_count)),
                ),
                instr::retc(),
                instr::label(suspended_get),
                instr::retc_suspended(),
            ])
        } else {
            InstrSeq::gather(vec![
                instr::memoget(notfound.clone(), Some((first_local.clone(), param_count))),
                instr::retc(),
            ])
        },
        instr::label(notfound),
        instr::nulluninit(),
        instr::nulluninit(),
        emit_memoize_helpers::param_code_gets(hhas_params),
        reified_get,
        instr::fcallfuncd(fcall_args, renamed_id),
        instr::memoset(Some((first_local.clone(), param_count))),
        if is_async {
            InstrSeq::gather(vec![
                instr::retc_suspended(),
                instr::label(eager_set),
                instr::memoset_eager(Some((first_local, param_count))),
                instr::retc(),
            ])
        } else {
            InstrSeq::gather(vec![instr::retc()])
        },
        default_value_setters,
    ]))
}

fn make_memoize_function_no_params_code(
    e: &mut Emitter,
    env: &mut Env,
    deprecation_info: Option<&[TypedValue]>,
    renamed_id: FunId<'static>,
    is_async: bool,
) -> Result {
    let notfound = e.label_gen_mut().next_regular();
    let suspended_get = e.label_gen_mut().next_regular();
    let eager_set = e.label_gen_mut().next_regular();
    let deprecation_body =
        emit_body::emit_deprecation_info(&env.scope, deprecation_info, e.systemlib())?;
    let fcall_args = FcallArgs::new(
        FcallFlags::default(),
        1,
        vec![],
        if is_async {
            Some(eager_set.clone())
        } else {
            None
        },
        0,
        None,
    );
    Ok(InstrSeq::gather(vec![
        deprecation_body,
        if is_async {
            InstrSeq::gather(vec![
                instr::memoget_eager(notfound.clone(), suspended_get.clone(), None),
                instr::retc(),
                instr::label(suspended_get),
                instr::retc_suspended(),
            ])
        } else {
            InstrSeq::gather(vec![instr::memoget(notfound.clone(), None), instr::retc()])
        },
        instr::label(notfound),
        instr::nulluninit(),
        instr::nulluninit(),
        instr::fcallfuncd(fcall_args, renamed_id),
        instr::memoset(None),
        if is_async {
            InstrSeq::gather(vec![
                instr::retc_suspended(),
                instr::label(eager_set),
                instr::memoset_eager(None),
                instr::retc(),
            ])
        } else {
            instr::retc()
        },
    ]))
}

fn make_wrapper_body<'a>(
    emitter: &mut Emitter,
    env: Env<'a>,
    return_type_info: HhasTypeInfo,
    params: Vec<HhasParam>,
    body_instrs: InstrSeq,
    is_reified: bool,
    has_coeffects_local: bool,
) -> Result<HhasBody> {
    let mut decl_vars = vec![];
    if is_reified {
        decl_vars.push(reified::GENERICS_LOCAL_NAME.into());
    }
    if has_coeffects_local {
        decl_vars.push(coeffects::LOCAL_NAME.into());
    }
    emit_body::make_body(
        emitter,
        body_instrs,
        decl_vars,
        true,   /* is_memoize_wrapper */
        false,  /* is_memoize_wrapper_lsb */
        vec![], /* upper_bounds */
        vec![], /* shadowed_tparams */
        params,
        Some(return_type_info),
        None, /* doc comment */
        Some(&env),
    )
}
